### PR TITLE
Ensure the 'Add' name actually matches before accepting any invocation expression.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/16158

### CODE REVIEW

File: src/Features/Core/Portable/UseCollectionInitializer/ObjectCreationExpressionAnalyzer.cs [hunk 1]
```diff
@@ -186,14 +186,15 @@ private void AddMatches(ArrayBuilder<TExpressionStatementSyntax> matches)
...
-            _syntaxFacts.GetPartsOfMemberAccessExpression(memberAccess, out instance, out var memberName);
+            _syntaxFacts.GetPartsOfMemberAccessExpression(memberAccess, out var localInstance, out var memberName);
             _syntaxFacts.GetNameAndArityOfSimpleName(memberName, out var name, out var arity);
 
             if (arity != 0 || !name.Equals(nameof(IList.Add)))
             {
                 return false;
             }
 
+            instance = localInstance;
```
Comment (Pilchie) [inline src/Features/Core/Portable/UseCollectionInitializer/ObjectCreationExpressionAnalyzer.cs]: How is this different?

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseCollectionInitializer/ObjectCreationExpressionAnalyzer.cs]: we were setting the value, but returning 'false'.  the caller used both the return value and the out value.  The intention was that if 'false' was returned, then null should be the out-val.

**Comment by CyrusNajmabadi** [comment]
@Pilchie Any other concerns?

**Comment by CyrusNajmabadi** [comment]
@srivatsn This fixes an issue where we offer to simplify a collection initializer, even if you weren't calling the 'Add' method that the collection initializer needs.

**Pull request merged**